slmigrate/mongohandler.py

```diff
@@ -25,39 +26,82 @@ def stop_mongo(proc):
     subprocess.run(cmd_to_run)
 
 
+def migrate_document(destination_collection, document):
+    try:
+        print("Migrating " + str(document['_id']))
+        destination_collection.insert_one(document)
+    except pyerr.DuplicateKeyError:
+        print("Docuement " + str(document['_id']) + " already exists. Skipping")
+
+
+def identify_metadata_conflict(destination_collection, source_document):
+    destination_query = {'$and': [{'workspace': source_document['workspace']}, {'path': source_document['path']}]}
+    destination_document = destination_collection.find_one(destination_query)
+    if destination_document:
+        return SimpleNamespace(**{'source_id': source_document['_id'], 'destination_id': destination_document['_id']})
+    else:
+        return None
+
+
+def merge_history_document(source_id, destination_id, destination_db):
+    destination_collection = destination_db.get_collection('values')
+    destination_collection.update_one({'metadataId': source_id}, {'$set': {'metadataId': destination_id}})
+
+
+def migrate_metadata_collection(source_db, destination_db):
+    collection_name = 'metadata'
+    source_collection = source_db.get_collection(collection_name)
+    source_collection_iterable = source_collection.find()
+    destination_collection = destination_db.get_collection(collection_name)
+    for document in source_collection_iterable:
+        conflict = identify_metadata_conflict(destination_collection, document)
+        if conflict:
+            print("Conflict Found! " + "source_id=" + str(conflict.source_id) + " destination_id=" + str(conflict.destination_id))
+            merge_history_document(conflict.source_id, conflict.destination_id, destination_db)
+        else:
+            migrate_document(destination_collection, document)
+
+
+def migrate_values_collection(source_db, destination_db):
+    collection_name = 'values'
+    collection_iterable = source_db.get_collection(collection_name).find()
+    destination_collection = destination_db.get_collection(collection_name)
+    for document in collection_iterable:
+        migrate_document(destination_collection, document)
+
+
+def check_merge_history_readiness(destination_db):
+    # look for fields that should be set when Org modeling is present. If they are missing exit.
+    collection_name = 'metadata'
+    destination_collection = destination_db.get_collection(collection_name)
+    if destination_collection.find({'workspace': {'$exists': False}}).count() > 0:
+        print("Database is not ready for migration. Update the connection string in C:\\ProgramData\\National Instruments\\Skyline\\Config\\TagHistorian.json to point to the nitaghistorian database in your MongoDB instance and restart Service Manager. Please see TODODLINK HERE for more detail")
```

Comment:
Is this using the absence of the workspace property to mean that the user is using a custom connection string?

Could it also be that the user is trying to run the migration on a system that has not been upgraded/does not have a problem yet? Or is there something else that can catch that?

Reply:
> Is this using the absence of the workspace property to mean that the user is using a custom connection string?

So in order to determine if the metadata in the "wrong" database matches the database in the "correct" database I need to compare both their workspace and the path (the two of these together define a unique tag). Given the nature of this bug the "correct" database isn't updated by the tag historian service on startup to include the workspace field. I'm asking the user here to update their connection string so that this startup step runs on the "correct" database and therefore the workspace field becomes available. This code should ensure if a customer created a tag in a workspace other than _defualt_ will be migrated correctly, and we don't attempt to merge it with tag history data in _default_ (in the case where the tag has the same path as one in the _default_ workspace. 

> Could it also be that the user is trying to run the migration on a system that has not been upgraded/does not have a problem yet? Or is there something else that can catch that?

Per doing this on a system that hasn't yet been upgraded/does have this problem yet: I need to do additional work to update the `capture_migration` and `restore_migration` functions to ensure they work when dealing with a remote mongodb. I will look into if we can detect being in that case and avoid this problem by setting the correct connection string in TagHistorian.json prior to the service starting back up. I will capture this todo and update these function in a future branch/pr.